### PR TITLE
[OptionValue] Translation System

### PR DESCRIPTION
**Pull request by vikey89**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

This PR is based from [#3975](https://github.com/Sylius/Sylius/issues/3975)

### CODE REVIEW

File: src/Sylius/Component/Variation/Model/OptionValue.php [hunk 1]
```diff
@@ -108,19 +124,17 @@ public function getName()
...
+    public function getPresentationOption()
     {
         if (null === $this->option) {
             throw new \BadMethodCallException('The option have not been created yet so you cannot access proxy methods.');
         }
-
         return $this->option->getPresentation();
     }
-}
+}
```
Comment (tristanbes) [inline src/Sylius/Component/Variation/Model/OptionValue.php]: missing eof new line

File: src/Sylius/Component/Variation/Model/OptionValueInterface.php [hunk 2]
```diff
@@ -53,9 +46,16 @@ public function setValue($value);
...
     public function getPresentation();
-}
+
+    /**
+     * Proxy method to access the presentation of real option object.
+     *
+     * @return string The presentation of object
+     */
+    public function getPresentationOption();
+}
```
Comment (tristanbes) [inline src/Sylius/Component/Variation/Model/OptionValueInterface.php]: missing eof new line

File: src/Sylius/Component/Variation/Model/OptionValueTranslation.php [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function setPresentation($presentation)
+    {
+        $this->presentation = $presentation;
+    }
+}
```
Comment (tristanbes) [inline src/Sylius/Component/Variation/Model/OptionValueTranslation.php]: missing eof new line

File: src/Sylius/Component/Variation/Model/OptionValueTranslationInterface.php [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+     */
+    public function getPresentation();
+
+    /**
+     * Set presentation.
+     *
+     * @param string $presentation
+     */
+    public function setPresentation($presentation);
+}
```
Comment (tristanbes) [inline src/Sylius/Component/Variation/Model/OptionValueTranslationInterface.php]: missing eof new line

File: src/Sylius/Component/Product/Model/OptionValueTranslationInterface.php [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+namespace Sylius\Component\Product\Model;
+
+use Sylius\Component\Variation\Model\OptionTranslationInterface as BaseOptionTranslationInterface;
+
+/**
+ * @author Vincenzo Provenza <vincenzo.provenza89@gmail.com>
+ */
+interface OptionValueTranslationInterface extends BaseOptionTranslationInterface
+{
+}
```
Comment (tristanbes) [inline src/Sylius/Component/Product/Model/OptionValueTranslationInterface.php]: missing eof new line

File: src/Sylius/Bundle/VariationBundle/Form/Type/OptionValueTranslationType.php [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getName()
+    {
+        return sprintf('sylius_%s_option_value_translation', $this->variableName);
+    }
+}
```
Comment (tristanbes) [inline src/Sylius/Bundle/VariationBundle/Form/Type/OptionValueTranslationType.php]: missing eof new line

File: src/Sylius/Bundle/VariationBundle/DependencyInjection/Configuration.php [hunk 7]
```diff
@@ -202,6 +205,43 @@ private function addClassesSection(ArrayNodeDefinition $node)
...
+                                            ->variableNode('option_value')->end()
+                                            ->arrayNode('classes')
+                                                ->isRequired()
+                                                ->addDefaultsIfNotSet()
+                                                ->children()
+                                                    ->scalarNode('model')->defaultValue(OptionValueTranslation::class)->cannotBeEmpty()->end()
+                                                    ->scalarNode('interface')->defaultValue(OptionValueTranslationInterface::class)->cannotBeEmpty()->end()
+                                                    ->scalarNode('controller')->defaultValue(ResourceController::class)->cannotBeEmpty()->end()
+                                                    ->scalarNode('repository')->cannotBeEmpty()->end()
+                                                    ->scalarNode('factory')->defaultValue(Factory::class)->end()
```
Comment (lchrusciel) [inline src/Sylius/Bundle/VariationBundle/DependencyInjection/Configuration.php]: Class factory should not be empty too `->cannotBeEmpty()`

File: src/Sylius/Bundle/VariationBundle/Form/Type/OptionValueTranslationType.php [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+{
+    /**
+     * @var string
+     */
+    protected $variableName;
+
+    /**
+     * @param string $dataClass
+     * @param array  $validationGroups
+     * @param        $variableName
```
Comment (lchrusciel) [inline src/Sylius/Bundle/VariationBundle/Form/Type/OptionValueTranslationType.php]: Missing typehint

File: src/Sylius/Bundle/VariationBundle/Form/Type/OptionValueTranslationType.php [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+class OptionValueTranslationType extends AbstractResourceType
+{
+    /**
+     * @var string
+     */
+    protected $variableName;
+
+    /**
+     * @param string $dataClass
+     * @param array  $validationGroups
```
Comment (lchrusciel) [inline src/Sylius/Bundle/VariationBundle/Form/Type/OptionValueTranslationType.php]: We are not aligning parameters vertically anymore

File: src/Sylius/Bundle/ProductBundle/Resources/config/doctrine/model/OptionValueTranslation.orm.xml [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+ For the full copyright and license information, please view the LICENSE
+ file that was distributed with this source code.
+
+-->
+
+<doctrine-mapping xmlns="http://doctrine-project.org/schemas/orm/doctrine-mapping">
+
+    <mapped-superclass name="Sylius\Component\Product\Model\OptionValueTranslation" table="sylius_product_option_value_translation" />
+
+</doctrine-mapping>
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ProductBundle/Resources/config/doctrine/model/OptionValueTranslation.orm.xml]: Missing blank line at the end of file

File: src/Sylius/Bundle/VariationBundle/Resources/config/doctrine/model/OptionValueTranslation.orm.xml [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+    <mapped-superclass name="Sylius\Component\Variation\Model\OptionValueTranslation"
+                       table="sylius_option_value_translation">
+        <id name="id" column="id" type="integer">
+            <generator strategy="AUTO" />
+        </id>
+
+        <field name="presentation" column="presentation" type="string" />
+    </mapped-superclass>
+
+</doctrine-mapping>
```
Comment (lchrusciel) [inline src/Sylius/Bundle/VariationBundle/Resources/config/doctrine/model/OptionValueTranslation.orm.xml]: Missing blank line

File: src/Sylius/Bundle/VariationBundle/Resources/config/validation.xml [hunk 12]
```diff
@@ -88,12 +88,25 @@
...
+                <option name="minMessage">sylius.option.presentation.min_length</option>
+                <option name="max">255</option>
+                <option name="maxMessage">sylius.option.presentation.max_length</option>
                 <option name="groups">sylius</option>
             </constraint>
         </property>
     </class>
 
-</constraint-mapping>
+</constraint-mapping>
```
Comment (lchrusciel) [inline src/Sylius/Bundle/VariationBundle/Resources/config/validation.xml]: again

File: src/Sylius/Component/Variation/Model/OptionValue.php [hunk 13]
```diff
@@ -10,11 +10,12 @@
  */
 
 namespace Sylius\Component\Variation\Model;
+use Sylius\Component\Translation\Model\AbstractTranslatable;
```
Comment (lchrusciel) [inline src/Sylius/Component/Variation/Model/OptionValue.php]: Missing blank line between namespace and use statement

File: src/Sylius/Component/Variation/Model/OptionValue.php [hunk 14]
```diff
@@ -37,11 +38,18 @@ class OptionValue implements OptionValueInterface
     protected $option;
 
     /**
+     * Displayed to user.
```
Comment (lchrusciel) [inline src/Sylius/Component/Variation/Model/OptionValue.php]: Isn't it redundant? Please, remove it

File: src/Sylius/Component/Variation/Model/OptionValueInterface.php [hunk 15]
```diff
@@ -3,7 +3,7 @@
 /*
  * This file is part of the Sylius package.
  *
- * (c) Paweł Jędrzejewski
+ * (c); Paweł Jędrzejewski
```
Comment (lchrusciel) [inline src/Sylius/Component/Variation/Model/OptionValueInterface.php]: Revert this change

File: src/Sylius/Component/Variation/Model/OptionValueTranslationInterface.php [hunk 16]
```diff
@@ -0,0 +1,34 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c); Paweł Jędrzejewski
```
Comment (lchrusciel) [inline src/Sylius/Component/Variation/Model/OptionValueTranslationInterface.php]: I guess you have something wrong with your templates. This header shouldn't contain semicolon

File: src/Sylius/Component/Variation/Model/OptionValueInterface.php [hunk 17]
```diff
@@ -11,13 +11,13 @@
...
 
+use Sylius\Component\Product\Model\OptionValueTranslationInterface;
 use Sylius\Component\Resource\Model\CodeAwareInterface;
-use Sylius\Component\Resource\Model\ResourceInterface;
 
 /**
  * @author Paweł Jędrzejewski <pawel@sylius.org>
  */
-interface OptionValueInterface extends ResourceInterface, CodeAwareInterface
+interface OptionValueInterface extends CodeAwareInterface, OptionValueTranslationInterface
```
Comment (lchrusciel) [inline src/Sylius/Component/Variation/Model/OptionValueInterface.php]: This change should be reverted. You shouldn't couple `OptionValueInterface` with `OptionValueTranslationInterface`. You can check how options are implemented to get a clue.

File: src/Sylius/Bundle/VariationBundle/DependencyInjection/Configuration.php (hunk 7 above)
Comment (lchrusciel) [inline src/Sylius/Bundle/VariationBundle/DependencyInjection/Configuration.php]: Another thing is that we should use **TranslatableFactory**, shouldn't we?

File: src/Sylius/Bundle/VariationBundle/Form/Type/OptionValueType.php [hunk 18]
```diff
@@ -43,11 +43,11 @@ public function __construct($dataClass, array $validationGroups, $variableName)
...
         $builder
-            ->add('value', 'text', [
-                'label' => 'sylius.form.option_value.value',
+            ->add('translations', 'a2lix_translationsForms', [
+                'form_type' => sprintf('sylius_%s_option_value_translation', $this->variableName),
+                'label' => 'sylius.form.option.presentation',
             ])
-            ->addEventSubscriber(new AddCodeFormSubscriber())
-        ;
+            ->addEventSubscriber(new AddCodeFormSubscriber());
```
Comment (lchrusciel) [inline src/Sylius/Bundle/VariationBundle/Form/Type/OptionValueType.php]: I am pretty sure we have convention to leave semicolon in a new line aligned to builder variable call.

**Comment by vikey89** [comment]
@pjedrzejewski I've finished on this PR. When you think you will merge this PR?
Thanks.

**Comment by lchrusciel** [comment]
@vikey89 Can you apply my 2 last [comments](https://github.com/Sylius/Sylius/pull/4051#discussion_r52141241)?

File: src/Sylius/Bundle/VariationBundle/DependencyInjection/Configuration.php (hunk 7 above)
Comment (pjedrzejewski) [inline src/Sylius/Bundle/VariationBundle/DependencyInjection/Configuration.php]: Correct about ``cannotBeEmpty()``, but Translation resources should have normal factory, only translatable resources use the Translatable one.

Comment (lchrusciel) [inline src/Sylius/Bundle/VariationBundle/DependencyInjection/Configuration.php]: :+1:

File: src/Sylius/Bundle/VariationBundle/DependencyInjection/Configuration.php [hunk 19]
```diff
@@ -184,7 +187,7 @@ private function addClassesSection(ArrayNodeDefinition $node)
                                             ->scalarNode('interface')->isRequired()->cannotBeEmpty()->end()
                                             ->scalarNode('controller')->defaultValue(ResourceController::class)->cannotBeEmpty()->end()
                                             ->scalarNode('repository')->cannotBeEmpty()->cannotBeEmpty()->end()
-                                            ->scalarNode('factory')->defaultValue(Factory::class)->end()
+                                            ->scalarNode('factory')->defaultValue(TranslatableFactory::class)->end()
```
Comment (lchrusciel) [inline src/Sylius/Bundle/VariationBundle/DependencyInjection/Configuration.php]: This node should cannot be empty as well

File: src/Sylius/Bundle/VariationBundle/Form/Type/OptionValueTranslationType.php [hunk 20]
```diff
@@ -0,0 +1,56 @@
...
+{
+    /**
+     * @var string
+     */
+    protected $variableName;
+
+    /**
+     * @param string $dataClass
+     * @param array $validationGroups
+     * @param $variableName
```
Comment (lchrusciel) [inline src/Sylius/Bundle/VariationBundle/Form/Type/OptionValueTranslationType.php]: Missing typehint

File: src/Sylius/Bundle/VariationBundle/Form/Type/OptionValueTranslationType.php [hunk 21]
```diff
@@ -0,0 +1,56 @@
...
+
+    /**
+     * {@inheritdoc}
+     */
+    public function buildForm(FormBuilderInterface $builder, array $options)
+    {
+        $builder
+            ->add('value', 'text', [
+                'label' => 'sylius.form.option_value.value',
+            ]);
```
Comment (Zales0123) [inline src/Sylius/Bundle/VariationBundle/Form/Type/OptionValueTranslationType.php]: According to cs this semicolon should be moved to next line.

File: src/Sylius/Component/Variation/Model/OptionValue.php [hunk 22]
```diff
@@ -108,7 +110,6 @@ public function getName()
         if (null === $this->option) {
             throw new \BadMethodCallException('The option have not been created yet so you cannot access proxy methods.');
         }
-
```
Comment (lchrusciel) [inline src/Sylius/Component/Variation/Model/OptionValue.php]: We should leave blank line before return statement.

File: src/Sylius/Component/Variation/Model/OptionValueTranslation.php [hunk 23]
```diff
@@ -0,0 +1,55 @@
...
+    /**
+     * @return string
+     */
+    public function getValue()
+    {
+        return $this->value;
+    }
+
+    /**
+     * @param $value
```
Comment (Zales0123) [inline src/Sylius/Component/Variation/Model/OptionValueTranslation.php]: Missing typehint.

Comment (Zales0123) [inline src/Sylius/Component/Variation/Model/OptionValueTranslation.php]: Anyway, it should be ``{@inheritdoc}``here and above ;)

File: src/Sylius/Component/Variation/Model/OptionValueTranslationInterface.php [hunk 24]
```diff
@@ -0,0 +1,34 @@
...
+{
+    /**
+     * The name displayed to user.
+     *
+     * @return string
+     */
+    public function getValue();
+
+    /**
+     * Set value.
```
Comment (Zales0123) [inline src/Sylius/Component/Variation/Model/OptionValueTranslationInterface.php]: Redundant comment.

File: src/Sylius/Bundle/ProductBundle/DependencyInjection/SyliusProductExtension.php [hunk 25]
```diff
@@ -151,6 +153,12 @@ private function prependVariation(ContainerBuilder $container, array $config)
                             'model' => OptionValue::class,
                             'interface' => OptionValueInterface::class,
                         ],
+                        'translation' => [
+                            'classes' => [
+                                'model' => OptionValueTranslation::class,
+                                'interface' => OptionValueTranslationInterface::class
```
Comment (Zales0123) [inline src/Sylius/Bundle/ProductBundle/DependencyInjection/SyliusProductExtension.php]: Missing comma at the end of line.

File: src/Sylius/Bundle/FixturesBundle/DataFixtures/ORM/LoadProductOptionData.php [hunk 26]
```diff
@@ -76,9 +76,9 @@ public function load(ObjectManager $manager)
             'Mug type',
             [$this->defaultLocale => 'Type', 'es_ES' => 'Tipo'],
             [
-                'OV12' => 'Medium mug',
-                'OV13' => 'Double mug',
-                'OV14' => 'MONSTER mug',
+                'OV12' => [$this->defaultLocale => 'Medium mug', 'es_ES' => 'Taza mediana'],
+                'OV13' => [$this->defaultLocale => 'Double mug', 'es_ES' => 'Taza doble'],
+                'OV14' => [$this->defaultLocale => 'MONSTER mug', 'es_ES' => 'Taza del monstruo']
```
Comment (lchrusciel) [inline src/Sylius/Bundle/FixturesBundle/DataFixtures/ORM/LoadProductOptionData.php]: Missing comma at the end of line

File: src/Sylius/Bundle/VariationBundle/Form/Type/OptionValueTranslationType.php [hunk 27]
```diff
@@ -0,0 +1,56 @@
...
+
+/**
+ * @author Vincenzo Provenza <vincenzo.provenza89@gmail.com>
+ */
+class OptionValueTranslationType extends AbstractResourceType
+{
+    /**
+     * @var string
+     */
+    protected $variableName;
```
Comment (Zales0123) [inline src/Sylius/Bundle/VariationBundle/Form/Type/OptionValueTranslationType.php]: I think that according to convention in all *Sylius*, it should be ``$subject``.

File: src/Sylius/Bundle/ProductBundle/DependencyInjection/SyliusProductExtension.php (hunk 25 above)
Comment (Zales0123) [inline src/Sylius/Bundle/ProductBundle/DependencyInjection/SyliusProductExtension.php]: And in two lines below ;)

File: src/Sylius/Bundle/VariationBundle/Form/Type/OptionValueTranslationType.php (hunk 27 above)
Comment (vikey89) [inline src/Sylius/Bundle/VariationBundle/Form/Type/OptionValueTranslationType.php]: I saw `$variableName` from [OptionTranslationType.php](https://github.com/Sylius/Sylius/blob/master/src/Sylius/Bundle/VariationBundle/Form/Type/OptionTranslationType.php#L25)

File: src/Sylius/Bundle/VariationBundle/DependencyInjection/Configuration.php (hunk 7 above)
Comment (vikey89) [inline src/Sylius/Bundle/VariationBundle/DependencyInjection/Configuration.php]: @pjedrzejewski @lchrusciel  Why should  be `->cannotBeEmpty()`? In [Option Node Configuration](https://github.com/Sylius/Sylius/blob/master/src/Sylius/Bundle/VariationBundle/DependencyInjection/Configuration.php#L116) the factory hasn't  `->cannotBeEmpty()` and has `TranslatableFactory` and not normal factory because it is translatable resources. Even `optionValue` is translatable resources.

It's correct?

Comment (lchrusciel) [inline src/Sylius/Bundle/VariationBundle/DependencyInjection/Configuration.php]: Configuration of [option node](https://github.com/Sylius/Sylius/blob/master/src/Sylius/Bundle/VariationBundle/DependencyInjection/Configuration.php#L116) is wrong. With refactoring we forgot to add this statement and we didn't catch it during review. It is not really bad mistake, but we try to fix it anyway. Thats why every new resources should have this option. 
Another thing is TranslatableFactory. You are correct, that option node has different factory, but there you have **translatable resource**. Here, you have to specify factory for **translation**. Do you understand the difference? If you check few lines below configuration you've referenced, you will see [translation configuration](https://github.com/Sylius/Sylius/blob/master/src/Sylius/Bundle/VariationBundle/DependencyInjection/Configuration.php#L149) which contains regular factory class.

Comment (vikey89) [inline src/Sylius/Bundle/VariationBundle/DependencyInjection/Configuration.php]: ok, thanks. I fixed in this [commit](https://github.com/vikey89/Sylius/commit/fc55513bb4608d721ae9aefac8f82724eb908c3d.).

**Comment by pjedrzejewski** [comment]
Nice work Vincenzo! Thank you very much. :+1:

**Comment by vikey89** [comment]
Thanks @pjedrzejewski I would like to help with something else :)
Any suggestion?

**Comment by michalmarcinkowski** [comment]
@vikey89 how about [this](https://github.com/Sylius/Sylius/issues/4213)? :)

**Comment by vikey89** [comment]
ok @michalmarcinkowski thanks :)
